avr-hal-generic/src/i2c.rs

```diff
@@ -599,73 +593,137 @@ macro_rules! impl_twi_i2c {
-                self.p.$twar.write(|w| unsafe {w.bits(self.address)});
-                self.p.$twcr.write(|w| w
-                    .$twen().set_bit()
-                    .$twea().set_bit()
-                    .$twsta().clear_bit()
-                    .$twsto().clear_bit()
-                    .$twint().set_bit()
-                );
+
+        // Slave State Machine
+        //
+        // Idea cribbed from [lpc8xx-hal](https://github.com/lpc-rs/lpc8xx-hal/blob/4205de2cdb0d54ab17b6f8d37aab34c518b606a6/src/i2c/slave.rs#L112)
+        //
+        // Valid Transitions are:
+        //
+        //      [<$I2c Slave>].init()
+        //          -> Initialized
+        //              -> [AddressMatched | Error]
+        //                  -> [RxReady | TxReady | Error]
+        //                      -> [Initialized | Error]
+        //
+        //
+        // # Example
+        //
+        // ```rust
+        // 
+        // ```
+        pub enum [<$I2c SlaveState>] {
+            /// Intialized state machine
+            Initialized([<$I2c SlaveStateInitialized>]),
+
+            /// Address sent by master has been matched
+            AddressMatched([<$I2c SlaveStateAddressMatched>]),
+
+            /// Data has been received from master
+            RxReady([<$I2c SlaveStateRxReady>]),
+
+            /// Ready to transmit data to master
+            TxReady([<$I2c SlaveStateTxReady>]),
+
+            /// Error State
+            Error([<$I2c SlaveStateError>]),
+        }
+
+       pub struct [<$I2c SlaveStateInitialized>];
+
+       impl [<$I2c SlaveStateInitialized>] {
+            pub fn wait() -> [<$I2c SlaveState>] {
+                // returns [<$I2c SlaveStateAddressMatched>] | [<$I2c SlaveStateError>]
+                [<$I2c SlaveState>]::Error([<$I2c SlaveStateError>])
             }
+        }
 
-//            fn wait(&mut self) {
+       pub struct [<$I2c SlaveStateAddressMatched>];
+
+       impl [<$I2c SlaveStateAddressMatched>]{
+           pub fn process() -> [<$I2c SlaveState>] {
+                // returns [<$I2c SlaveStateRxReady>] | [<$I2c SlaveStateTxReady>] | [<$I2c SlaveStateError>]
+                [<$I2c SlaveState>]::Error([<$I2c SlaveStateError>])
+           }
+       }
+
+       pub struct [<$I2c SlaveStateRxReady>]{
+           data: u8,
+       }
+
+       impl [<$I2c SlaveStateRxReady>]{
+           pub fn read() -> [<$I2c SlaveState>] {
+                // returns [<$I2c SlaveStateRxReady>] | [<$I2c SlaveStateError>]
+                [<$I2c SlaveState>]::Error([<$I2c SlaveStateError>])
+           }
+       }
+       pub struct [<$I2c SlaveStateTxReady>];
+
+       impl [<$I2c SlaveStateTxReady>]{
+           pub fn write(_: u8) -> [<$I2c SlaveState>] {
+                // returns [<$I2c SlaveStateTxReady>] | [<$I2c SlaveStateError>]
+                [<$I2c SlaveState>]::Error([<$I2c SlaveStateError>])
+           }
+       }
+       pub struct [<$I2c SlaveStateError>];
+
+
+//       impl [<$I2c Slave>]
```

Comment:
Couple ways to go here...

    Slave.init() -> the SlaveState::Initialized
 
or

    SlaveState::Unitialized.init(slave) -> SlaveState::Initialized